dnw-1.0.0/doc/dnw.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DNW 1"
.TH DNW 1 "2011-03-09" "dnw-1.0.0" "Samsung USB Bulk Transfer Utility"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dnw \- Samsung USB Bulk Transfer Utility
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdnw\fR [\fI\s-1OPTION\s0\fR]... \fI\s-1FILE\s0\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBdnw\fR is a \s-1USB\s0 bulk transfer utility for Samsung devices.
.PP
The purpose of this program is to transfer firmware files to a Samsung
electronic device when this one is placed in an appropriate recovery or
boot mode.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fBdnw\fR accepts the following command line options:
.IP "\fB\-a\fR \fI\s-1ADDRESS\s0\fR, \fB\-A\fR \fI\s-1ADDRESS\s0\fR, \fB\-\-address=\fR\fI\s-1ADDRESS\s0\fR" 4
.IX Item "-a ADDRESS, -A ADDRESS, --address=ADDRESS"
Specify the download \s-1ADDRESS\s0 (default is c0008000). Note that this
address may be ignored by the target device.
.IP "\fB\-b\fR \fI\s-1SIZE\s0\fR, \fB\-B\fR \fI\s-1SIZE\s0\fR, \fB\-\-block\-size=\fR\fI\s-1SIZE\s0\fR" 4
.IX Item "-b SIZE, -B SIZE, --block-size=SIZE"
Specify the \s-1USB\s0 transfer block \s-1SIZE\s0 (default is 512).
.IP "\fB\-c\fR \fI\s-1VALUE\s0\fR, \fB\-C\fR \fI\s-1VALUE\s0\fR, \fB\-\-config=\fR\fI\s-1VALUE\s0\fR" 4
.IX Item "-c VALUE, -C VALUE, --config=VALUE"
Specify the \s-1USB\s0 configuration \s-1VALUE\s0 (default is 1).
.IP "\fB\-d\fR \fI\s-1VENDOR:PRODUCT\s0\fR, \fB\-D\fR \fI\s-1VENDOR:PRODUCT\s0\fR, \fB\-\-device=\fR\fI\s-1VENDOR:PRODUCT\s0\fR" 4
.IX Item "-d VENDOR:PRODUCT, -D VENDOR:PRODUCT, --device=VENDOR:PRODUCT"
Specify the \s-1USB\s0 \s-1VENDOR:PRODUCT\s0 \s-1ID\s0 (default is \*(L"04E8:1234\*(R").
.IP "\fB\-e\fR \fI\s-1ADDRESS\s0\fR, \fB\-E\fR \fI\s-1ADDRESS\s0\fR, \fB\-\-endpoint=\fR\fI\s-1ADDRESS\s0\fR" 4
.IX Item "-e ADDRESS, -E ADDRESS, --endpoint=ADDRESS"
Specify the \s-1USB\s0 transfer bulk \s-1OUT\s0 endpoint \s-1ADDRESS\s0 (default is 2).
.IP "\fB\-h\fR, \fB\-H\fR, \fB\-\-help\fR" 4
.IX Item "-h, -H, --help"
Show summary of options.
.IP "\fB\-i\fR \fI\s-1NUMBER\s0\fR, \fB\-I\fR \fI\s-1NUMBER\s0\fR, \fB\-\-interface=\fR\fI\s-1NUMBER\s0\fR" 4
.IX Item "-i NUMBER, -I NUMBER, --interface=NUMBER"
Specify the \s-1USB\s0 interface \s-1NUMBER\s0 (default is 0).
.IP "\fB\-q\fR, \fB\-Q\fR, \fB\-\-quiet\fR" 4
.IX Item "-q, -Q, --quiet"
Turn off display messages.
.IP "\fB\-t\fR \fI\s-1DURATION\s0\fR, \fB\-T\fR \fI\s-1DURATION\s0\fR, \fB\-\-timeout=\fR\fI\s-1DURATION\s0\fR" 4
.IX Item "-t DURATION, -T DURATION, --timeout=DURATION"
Specify the \s-1USB\s0 transfer timeout \s-1DURATION\s0 in ms (default is 3000).
.IP "\fB\-v\fR, \fB\-V\fR, \fB\-\-version\fR" 4
.IX Item "-v, -V, --version"
Show version of program.
.PP
More information on the values to specify for the above parameters can
be found by using the \fIlsusb\fR\|(8) utility on the given device.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBlsusb\fR(8).
.SH "AUTHORS"
.IX Header "AUTHORS"
Michel Stempin <mailto://michel.stempin@wanadoo.fr>.
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
If you feel you have found a bug in \fBdnw\fR, please submit a bug
report at <mailto://michel.stempin@wanadoo.fr>.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2011 Michel Stempin <mailto://michel.stempin@wanadoo.fr>.
.PP
This program comes with \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by
law. You may redistribute copies of this program under the terms of
the \s-1GNU\s0 General Public License. For more information about these
matters, see the file named \s-1COPYING\s0.
.SH "DISCLAIMER"
.IX Header "DISCLAIMER"
\&\fBdnw\fR is provided ``\s-1AS\s0 \s-1IS\s0'' and any express or implied warranties,
including, but not limited to, the implied warranties of
merchantability and fitness for a particular purpose are disclaimed.
See the \s-1COPYING\s0 file distributed with \fBdnw\fR for complete details.


secbulk-1.0.0/NEWS

 secbulk - Samsung USB bulk-transfer Linux kernel driver
 Copyright (C) 2011 Michel Stempin <michel.stempin@wanadoo.fr>
 Heavily based on existing GPL source code that can be found in
 various Google Code projects and on several Chineese blog sites.

 $Id: NEWS 214 2011-03-09 23:36:14Z michel.stempin@wanadoo.fr $

================================================================================

Version 1.0.0

  * Initial version for Samsung SNE-60 eReader device


dnw-1.0.0/src/dnw.c
/*
 * dnw.c -- Samsung USB bulk-transfer utility
 *
 * Copyright (C) 2011 Michel Stempin <michel.stempin@wanadoo.fr>
 * Based on previous work by Fox <hulifox008@163.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions, and the following disclaimer,
 *    without modification.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The names of the above-listed copyright holders may not be used
 *    to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * ALTERNATIVELY, this software may be distributed under the terms of the
 * GNU General Public License ("GPL") as published by the Free Software
 * Foundation, either version 2 of that License or (at your option) any
 * later version.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <locale.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <usb.h>
#include <sys/stat.h>
#include "config.h"
#include "gettext.h"

// Define shortcut for gettext()
#define _(string) gettext(string)

#define DEFAULT_ADDRESS      (0xC0008000)
#define	DEFAULT_BLOCK_SIZE   (512)
#define	DEFAULT_PRODUCT      (0x1234)
#define DEFAULT_VENDOR       (0x04E8)
#define DEFAULT_CONFIG       (1)
#define DEFAULT_INTERFACE    (0)
#define DEFAULT_EP           (0x02)
#define DEFAULT_TIMEOUT      (3000)

//==============================================================================
// Private Data
//==============================================================================

// Download address
static u_int32_t address = DEFAULT_ADDRESS;

// Transfer block size
static int block_size = DEFAULT_BLOCK_SIZE;

// USB Product ID
static unsigned int product = DEFAULT_PRODUCT;

// USB Vendor ID
static unsigned int vendor = DEFAULT_VENDOR;

// USB configuration number
static int configuration = DEFAULT_CONFIG;

// USB interface number
static int interface = DEFAULT_INTERFACE;

// USB bulk transfer endpoint
static int endpoint = DEFAULT_EP;

// USB transfer timeout
static int timeout = DEFAULT_TIMEOUT;

// Quiet mode
static int quiet = 0;

// Source filename
static char *filename = NULL;

//==============================================================================
// Private functions
//==============================================================================

// Print usage information
static void
print_usage(void)
{
  printf(_("Usage: dnw [options] file\n"));
  printf(_("Options:\n"));
  printf(_(" -a, -A, --address=ADDRESS              Specify the download ADDRESS (default is 0xc0008000)\n"));
  printf(_(" -b, -B, --block-size=SIZE              Specify the USB transfer block SIZE (default is 512)\n"));
  printf(_(" -c, -C, --config=VALUE                 Specify the USB configuration VALUE (default is 1)\n"));
  printf(_(" -d, -D, --device=VENDOR:PRODUCT        Specify the USB VENDOR:PRODUCT device ID (default is \"04E8:1234\")\n"));
  printf(_(" -e, -E, --endpoint=ADDRESS             Specify the USB transfer endpoint ADDRESS (default is 2)\n"));
  printf(_(" -h, -H, --help                         Print the option help\n"));
  printf(_(" -i, -I, --interface=NUMBER             Specify the USB interface number (default is 0)\n"));
  printf(_(" -q, -Q, --quiet                        Turn off display messages\n"));
  printf(_(" -t, -T, --timeout=DURATION             Specify the USB transfer timemout DURATION in ms (default is 3000)\n"));
  printf(_(" -v, -V, --version                      Print the version information\n"));
}

// Print version information
static void
print_version(void)
{
  printf("dnw version "VERSION"\n");
  printf(_("USB bulk-transfer utility for Samsung devices.\n"));
  printf(_("Copyright (C) 2011, Michel Stempin <michel.stempin@wanadoo.fr>\n"));
  printf(_("License GPL2+: GNU GPL version 2 or later <http://www.gnu.org/licenses/gpl-2.0.html>.\n"));
  printf(_("This is free software: you are free to change and redistribute it.\n"));
  printf(_("There is NO WARANTY, to the extent permitted by law.\n"));
}

// Parse command line options
static void
parse_options(int argc, char *argv[])
{
  struct option long_options[] = {
    {"address", 1, NULL, 0},
    {"block-size", 1, NULL, 0},
    {"config", 1, NULL, 0},
    {"device", 1, NULL, 0},
    {"endpoint", 1, NULL, 0},
    {"help", 0, NULL, 0},
    {"interface", 1, NULL, 0},
    {"quiet", 0, NULL, 0},
    {"timeout", 1, NULL, 0},
    {"version", 0, NULL, 0},
    {0, 0, NULL, 0}
  };
  int c, opt;

  while (1) {
    c = getopt_long(argc, argv, "a:A:b:B:e:E:hHi:i:qQt:T:vV", long_options, &opt);
    if (c == -1) {
      break;
    }
    if (c == 0) {
      if (!strcmp(long_options[opt].name, "address")) {
        c = 'a';
      } else if (!strcmp(long_options[opt].name, "block-size")) {
        c = 'b';
      } else if (!strcmp(long_options[opt].name, "config")) {
        c = 'c';
      } else if (!strcmp(long_options[opt].name, "device")) {
        c = 'd';
      } else if (!strcmp(long_options[opt].name, "endpoint")) {
        c = 'e';
      } else if (!strcmp(long_options[opt].name, "help")) {
        c = 'h';
      } else if (!strcmp(long_options[opt].name, "interface")) {
        c = 'i';
      } else if (!strcmp(long_options[opt].name, "quiet")) {
        c = 'q';
      } else if (!strcmp(long_options[opt].name, "timeout")) {
        c = 't';
      } else if (!strcmp(long_options[opt].name, "version")) {
        c = 'v';
      }
    }
    switch (c) {
    case 'a':
    case 'A':
      if (optarg != NULL) {
        address = atoi(optarg);
      }
      break;
    case 'b':
    case 'B':
      if (optarg != NULL) {
        block_size = atoi(optarg);
      }
      break;
    case 'c':
    case 'C':
      if (optarg != NULL) {
        configuration = atoi(optarg);
      }
      break;
    case 'd':
    case 'D':
      if (optarg != NULL) {
        sscanf(optarg, "%x:%x", &vendor, &product);
      }
      break;
    case 'e':
    case 'E':
      if (optarg != NULL) {
        endpoint = atoi(optarg);
      }
      break;
    case 'h':
    case 'H':
      print_usage();
      exit(EXIT_SUCCESS);
    case 'i':
    case 'I':
      if (optarg != NULL) {
        interface = atoi(optarg);
      }
      break;
    case 'q':
    case 'Q':
      if (optarg != NULL) {
        quiet = atoi(optarg);
      }
      break;
    case 't':
    case 'T':
      if (optarg != NULL) {
        timeout = atoi(optarg);
      }
      break;
    case 'v':
    case 'V':
      print_version();
      exit(EXIT_SUCCESS);
    default:
      print_usage();
      exit(EXIT_FAILURE);
    }
  }
  if (optind >= argc) {
    print_usage();
    exit(EXIT_FAILURE);
  }
  filename = strdup(argv[optind]);
}

// Get a handle on the target USB device
static struct usb_dev_handle *
open_device(uint16_t vendor, uint16_t product, int configuration, int interface)
{
	struct usb_bus *busses, *bus;
  struct usb_device *device;
  struct usb_dev_handle *hd;

	usb_init();
	usb_find_busses();
	usb_find_devices();

  // Scan USB busses
	busses = usb_get_busses();
	for (bus = busses; bus; bus = bus->next) {

    // Scan USB devices
		for (device = bus->devices; device; device = device->next) {

      // Look for our device
			if (device->descriptor.idVendor == vendor &&
          device->descriptor.idProduct == product) {

        // Device found
        if (!quiet) {
          printf(_("Target USB device found!\n"));
        }
				hd = usb_open(device);
				if (!hd) {
					perror(_("Cannot open USB device"));
          return NULL;
				}
        if (usb_set_configuration(hd, configuration) < 0) {
          perror(_("Cannot set configuration for USB device"));
          usb_close(hd);
          return NULL;
        }
        if (usb_claim_interface(hd, interface) < 0) {
          perror(_("Cannot claim interface for USB device"));
          usb_close(hd);
          return NULL;
				}
				return hd;
			}
		}
	}
  fprintf(stderr, _("Target USB device not found!\n"));
	return NULL;
}

// Read the source file to transfer
unsigned char *read_file(char *filename, unsigned int *length)
{
	unsigned char *buffer = NULL;
	struct stat fs;
	int fd = -1, i;
	u_int16_t checksum = 0;

  // Open the file to transfer
	fd = open(filename, O_RDONLY);
	if (fd == -1) {
		perror(_("Cannot open file"));
		return NULL;
	}

  // Get the file size
	if (fstat(fd, &fs) == -1) {
		perror(_("Cannot get file size"));
		goto error;
	}

  // Allocate buffer
	buffer = (unsigned char *) malloc(fs.st_size + 10);
	if (buffer == NULL) {
		perror(_("Cannot allocate buffer memory"));
    goto error;
	}

  // Read the whole file into memory
  if (read(fd, buffer + 8, fs.st_size) != fs.st_size) {
		perror(_("Cannot read file"));
		goto error;
	}

  // Compute checksum
	for (i = 8; i < fs.st_size + 8; i++) {
		checksum += buffer[i];
	}
  if (!quiet) {
    printf(_("Filename : %s\n"), filename);
    printf(_("Filesize : %ld bytes\n"), fs.st_size);
    printf(_("Checksum : %#04x\n"), checksum);
  }

  // Transfer header: download address & size
	*((u_int32_t *) buffer) = address;
	*((u_int32_t *) buffer + 1) = fs.st_size + 10;

  // Transfer footer: checksum
  *((u_int16_t *) (buffer + fs.st_size + 8)) = checksum;

	*length = fs.st_size + 10;
	return buffer;

error:
	if (fd != -1) {
    close(fd);
  }
	if (buffer != NULL) {
    free(buffer);
  }
	return NULL;
	
}

//==============================================================================
// Public functions
//==============================================================================

// Main function
int
main(int argc, char *argv[])
{
	struct usb_dev_handle *device = NULL;
	unsigned int length = 0;
	unsigned char *buffer = NULL;
	unsigned int remain = 0;
	unsigned int to_write = 0;

  // Set up i18n
  setlocale(LC_ALL, "");
  textdomain("dnw");
  bindtextdomain("dnw", LOCALEDIR);

  // Parse command line options
  parse_options(argc, argv);

  // Open the USB device
	device = open_device(vendor, product, configuration, interface);
	if (!device) {
    goto error;
	}

  // Read the file into memory
	buffer = read_file(filename, &length);
	if (buffer == NULL) {
    goto error;
  }

  // Actually transfer the data
	remain = length;
  if (!quiet) {
    printf(_("Writing data...\n"));
  }
	while (remain) {
		to_write = (remain > block_size) ? block_size : remain; 
		if (usb_bulk_write(device,
                       endpoint,
                       (char *) buffer + (length - remain),
                       to_write,
                       timeout) != to_write) {
			perror(_("USB transfer failed"));
			goto error;
		}
		remain -= to_write;
    if (!quiet) {
      printf(_("\r%d%%\t %d bytes     "),
             (length - remain) * 100 / length,
             length - remain);
      fflush(stdout);
    }
	}
	if (!remain && !quiet) {
    printf(_("Done!\n"));
  }

  // Cleanup
  if (filename) {
    free(filename);
  }
  if (buffer) {
    free(buffer);
  }
  if (device) {
    usb_release_interface(device, 0);
    usb_close(device);
  }
	exit(EXIT_SUCCESS);

 error:
  if (filename) {
    free(filename);
  }
  if (buffer) {
    free(buffer);
  }
  if (device) {
    usb_release_interface(device, 0);
    usb_close(device);
  }
  exit(EXIT_FAILURE);
}
